File: src/main/java/no/fintlabs/provider/security/AdapterRegistrationValidator.java
package no.fintlabs.provider.security;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.fintlabs.adapter.models.AdapterCapability;
import no.fintlabs.provider.exception.InvalidAdapterCapabilityException;
import no.fintlabs.provider.security.resource.ResourceContext;
import org.springframework.stereotype.Component;

import java.util.Set;

@Slf4j
@Component
@RequiredArgsConstructor
public class AdapterRegistrationValidator {

    private final ResourceContext resourceContext;

    public void validateCapabilities(Set<AdapterCapability> capabilities) {
        capabilities.forEach(capability -> {
            String componentResource = "%s-%s-%s".formatted(capability.getDomainName(), capability.getPackageName(), capability.getResourceName()).toLowerCase();
            if (!resourceContext.getValidResources().contains(componentResource)) {
                log.warn("Validation failed: Capability '{}' from '{}' is not a valid resource.", capability, componentResource);
                throw new InvalidAdapterCapabilityException("Invalid capability resource: %s - Component does not exist".formatted(componentResource));
            } else if (invalidFullSyncInterval(capability.getFullSyncIntervalInDays())) {
                log.warn("Validation failed: Capability '{}' has an invalid FullSyncIntervalInDays value", capability.getEntityUri());
                throw new InvalidAdapterCapabilityException("Invalid capability resource: %s - FullSyncIntervalInDays value is invalid".formatted(componentResource));
            }
        });
    }

    private boolean invalidFullSyncInterval(int fullSyncIntervalInDays) {
        return fullSyncIntervalInDays <= 0;
    }

}

File: src/main/java/no/fintlabs/provider/security/AdapterRegistrationValidator.kt
package no.fintlabs.provider.security

import no.fintlabs.adapter.models.AdapterCapability
import no.fintlabs.adapter.models.AdapterContract
import no.fintlabs.provider.exception.InvalidAdapterCapabilityException
import no.fintlabs.provider.register.ContractJpaRepository
import no.fintlabs.provider.security.resource.ComponentResourceRegistry
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class AdapterRegistrationValidator(
    private val componentResourceRegistry: ComponentResourceRegistry,
    private val contractRepository: ContractJpaRepository
) {

    companion object {
        const val MAX_FULL_SYNC_INTERVAL_DAYS = 7
    }

    private val logger = LoggerFactory.getLogger(javaClass)

    fun validateContract(contract: AdapterContract) {
        val duplicates = duplicateResourceNamesInOrganisation(contract)
        contract.capabilities.forEach { capability ->
            rejectIfDuplicate(capability, contract.orgId, duplicates)
            rejectIfUnknownResource(capability)
            rejectIfInvalidFullSyncInterval(capability)
        }
    }

    private fun duplicateResourceNamesInOrganisation(contract: AdapterContract): Set<String> =
        contractRepository.findByOrgIdWithCapabilities(contract.orgId)
            .filter { it.userName != contract.username }
            .flatMap { it.capabilityEntityset }
            .mapTo(mutableSetOf()) { it.resourceName }

    private fun rejectIfDuplicate(capability: AdapterCapability, orgId: String, duplicates: Set<String>) {
        if (capability.resourceName !in duplicates) return
        logger.warn("Validation failed: Capability '$capability' from '${capability.entityUri}' is a duplicate in organisation '$orgId'")
        throw InvalidAdapterCapabilityException("Duplicate capability resource: ${capability.entityUri} - Organisation already has a capability with the same resource name")
    }

    private fun rejectIfUnknownResource(capability: AdapterCapability) {
        if (componentResourceRegistry.containsResource(
                capability.domainName,
                capability.packageName,
                capability.resourceName
            )
        ) return
        logger.warn("Validation failed: Capability '$capability' from '${capability.entityUri}' is not a valid resource.")
        throw InvalidAdapterCapabilityException("Invalid capability resource: ${capability.entityUri} - Component does not exist")
    }

    private fun rejectIfInvalidFullSyncInterval(capability: AdapterCapability) {
        if (capability.fullSyncIntervalInDays in 1..MAX_FULL_SYNC_INTERVAL_DAYS) return
        logger.warn("Validation failed: Capability '$capability' has an invalid FullSyncIntervalInDays value")
        throw InvalidAdapterCapabilityException("Invalid capability resource: ${capability.entityUri} - FullSyncIntervalInDays value is invalid")
    }

}